frame/vault/src/tests.rs

```diff
@@ -522,96 +532,100 @@ proptest! {
+// Liquidity Providers deposit native tokens into the vault
 proptest! {
 	#![proptest_config(ProptestConfig::with_cases(100))]
 
 	#[test]
 	fn vault_stock_dilution_k(
-		(random_index, created_accounts) in
+		(strategy_moment, accounts) in
 			valid_amounts_without_overflow_k_with_random_index(500, 1_000_000_000)
 				.prop_filter("a minimum of two accounts are required, 1 for the strategy and 1 depositor",
 							 |(_, x)| x.len() > 2)
 	) {
-		let asset_id = MockCurrencyId::D;
-		let (strategy_account_id, strategy_profits) = created_accounts[0];
-		let strategy_deposit_moment = random_index;
-		let account_start = 1;
-		let accounts = &created_accounts[account_start..];
+	let lps_start = 1;
+	let asset_id = MockCurrencyId::D;
 
-		ExtBuilder::default().build().execute_with(|| {
-			let (vault_id, vault) = create_vault(strategy_account_id, asset_id);
+	let (strategy_account_id, strategy_profits, strategy_losses) = accounts[0];
+	let lps = &accounts[lps_start..];
 
-			prop_assert_eq!(Tokens::balance(asset_id, &strategy_account_id), 0);
-			prop_assert_ok!(Tokens::mint_into(asset_id, &strategy_account_id, strategy_profits));
+	let strategy_diff = strategy_profits - strategy_losses;
+	let before_moment_lps = || lps.iter().take(strategy_moment).copied();
+	let after_moment_lps = || lps.iter().skip(strategy_moment).copied();
 
-			for (account, balance) in accounts.iter().copied() {
+	ExtBuilder::default().build().execute_with(|| {
+			let (vault_id, vault) = create_vault(strategy_account_id, asset_id);
+
+			// Mints native tokens for all accounts
+			for (account, initial_native_tokens, _) in accounts.iter().copied() {
 				prop_assert_eq!(Tokens::balance(asset_id, &account), 0);
-				prop_assert_ok!(Tokens::mint_into(asset_id, &account, balance));
+				prop_assert_ok!(Tokens::mint_into(asset_id, &account, initial_native_tokens));
 			}
 
-			// Shareholders
-			for (account, balance) in accounts[0..strategy_deposit_moment].iter().copied() {
-				prop_assert_ok!(
-					Vaults::deposit(Origin::signed(account), vault_id, balance)
-				);
+			// Liquidity providers deposit all their native tokens to receive LP tokens
+			// BEFORE losses and profits
+			for (account, initial_native_tokens, _) in before_moment_lps() {
+				let origin = Origin::signed(account);
+				prop_assert_ok!(Vaults::deposit(origin, vault_id, initial_native_tokens));
 			}
 
-			// Profits comming
-			prop_assert_ok!(
-				<Vaults as StrategicVault>::deposit(
-					&vault_id,
-					&strategy_account_id,
-					strategy_profits
-				)
-			);
-
-			// Shareholders total LP
-			let lp_total = Tokens::total_issuance(vault.lp_token_id);
+			prop_assert_ok!(<Vaults as StrategicVault>::deposit(
+				&vault_id,
+				&strategy_account_id,
+				strategy_profits
+			));
+			prop_assert_ok!(<Vaults as StrategicVault>::withdraw(
+				&vault_id,
+				&strategy_account_id,
+				strategy_losses
+			));
```

Comment:
It is hard for me to understand why `update_strategy_report` should be used because `withdraw` already modifies storage.

https://github.com/ComposableFi/composable/blob/823aa161930734af1a4689facd50b1b128704a51/frame/vault/src/lib.rs#L838-L839

Does the fact that currencies are being transferred makes testing losses impractical? For example, LPs won't be able to withdrawn LP tokens from the subtracted vault balances.

https://github.com/ComposableFi/composable/blob/823aa161930734af1a4689facd50b1b128704a51/frame/vault/src/lib.rs#L845-L846

Reply:
> 

`StrategicVault::withdraw` does indeed mutate the balance, but `StrategicVault::deposit` subtract the amount transferred back from the strategy balance. Hence, if you `StrategicVault::withdraw` for an amount `X` and then `ReportableStrategicVault::update_strategy_report` with an amount `Y` which is `< X`, you effectively declare a **loss**. Because `https://github.com/ComposableFi/composable/blob/551e705396608de1d469e8836c87bb7e731fed8c/frame/vault/src/lib.rs#L869` will be lowered (the difference `X - Y` has not been transferred back to the vault account).

Reply:
> `wittdraw(X)` and then `update_strategy_report(Y)` with `X` > `Y`

In a withdraw, the strategy vault balance will have a value `Y` lesser than the transferred amount `X` executed on `withdraw` because of `update_strategy_report`. Therefore, the strategy account balance will be greater than the strategy vault balance meaning a **loss**.

Yes, it makes sense locally but not in an overall architecture picture but such knowledge will eventually come with more development.